src/NzbDrone.Core/Download/History/DownloadHistoryService.cs

```diff
@@ -0,0 +1,164 @@
+﻿using System;
+using NzbDrone.Common.Extensions;
+using NzbDrone.Core.History;
+using NzbDrone.Core.MediaFiles.Events;
+using NzbDrone.Core.Messaging.Events;
+using NzbDrone.Core.Tv.Events;
+
+namespace NzbDrone.Core.Download.History
+{
+    public interface IDownloadHistoryService
+    {
+        bool DownloadAlreadyImported(string downloadId);
+        DownloadHistory GetLatestDownloadHistoryItem(string downloadId);
+    }
+
+    public class DownloadHistoryService : IDownloadHistoryService,
+                                          IHandle<EpisodeGrabbedEvent>,
+                                          IHandle<EpisodeImportedEvent>,
+                                          IHandle<DownloadCompletedEvent>,
+                                          IHandle<DownloadFailedEvent>,
+                                          IHandle<DownloadIgnoredEvent>,
+                                          IHandle<SeriesDeletedEvent>
+
+    {
+        private readonly IDownloadHistoryRepository _repository;
+        private readonly IHistoryService _historyService;
+
+        public DownloadHistoryService(IDownloadHistoryRepository repository, IHistoryService historyService)
+        {
+            _repository = repository;
+            _historyService = historyService;
+        }
+
+        public bool DownloadAlreadyImported(string downloadId)
+        {
+            var events = _repository.FindByDownloadId(downloadId);
+
+            // Events are ordered by date descending, if a grabbed event comes before an imported event then it was never imported
+            // or grabbed again after importing and should be reprocessed.
+            foreach (var e in events)
+            {
+                if (e.EventType == DownloadHistoryEventType.DownloadGrabbed)
+                {
+                    return false;
+                }
+
+                if (e.EventType == DownloadHistoryEventType.DownloadImported)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public DownloadHistory GetLatestDownloadHistoryItem(string downloadId)
+        {
+            var events = _repository.FindByDownloadId(downloadId);
+
+            // Events are ordered by date descending. We'll return the most recent expected event.
+            foreach (var e in events)
```

Comment:
What you're doing is just a `FirstOrDefault(e => e.EventType == .. || ..)`

Reply:
True, though I think this might be easier to grok and reduces the need to recheck the event type.